platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/RuntimeWorkspaceRegistry.java

```diff
@@ -334,54 +306,123 @@ public RuntimeWorkspaceImpl get(String workspaceId) throws NotFoundException {
+        final MachineConfig devCfg = findDev(machineConfigs);
+
+        // Starting workspace dev machine
+        final MachineImpl devMachine;
+        try {
+            devMachine = createMachine(devCfg, workspaceId, environment.getName(), recover);
+        } catch (MachineException | BadRequestException | NotFoundException | SnapshotException | ConflictException ex) {
+            doRemoveWorkspace(workspaceId);
+            throw ex;
         }
 
-        final List<MachineImpl> machines = new ArrayList<>();
+        // Updating workspace status and machines list
+        lock.writeLock().lock();
+        try {
+            addRunningMachine(devMachine);
+            final RuntimeWorkspaceImpl workspace = idToWorkspaces.get(workspaceId);
+            if (workspace != null) {
+                workspace.setStatus(RUNNING);
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
 
-        MachineConfig devMachine = findDev(environment.getMachineConfigs());
-        if (devMachine == null) {
-            throw new BadRequestException("Dev machine was not found in workspace environment " + environment.getName());
+        // Start all the other machines
+        machineConfigs.remove(devCfg);
+        for (MachineConfig nonDevCfg : machineConfigs) {
+            try {
+                addRunningMachine(createMachine(nonDevCfg, workspaceId, environment.getName(), recover));
+            } catch (MachineException | BadRequestException | NotFoundException | SnapshotException | ConflictException ex) {
+                LOG.error(format("Error while creating machine '%s' in workspace '%s', environment '%s'",
+                                 nonDevCfg.getName(),
+                                 workspaceId,
+                                 environment.getName()),
+                          ex);
+            }
         }
+    }
 
-        machines.add(createMachine(devMachine, workspaceId, environment.getName(), recover));
+    /**
+     * Checks if workspace is valid for start.
+     */
+    private void checkWorkspaceIsValidForStart(UsersWorkspace workspace, String envName) throws BadRequestException {
+        if (workspace == null) {
+            throw new BadRequestException("Required non-null workspace");
+        }
+        if (envName == null) {
+            throw new BadRequestException(format("Couldn't start workspace '%s', environment name is null", workspace.getName()));
+        }
+        final EnvironmentState environment = workspace.getEnvironments().get(envName);
+        if (environment == null) {
+            throw new BadRequestException(format("Couldn't start workspace '%s', workspace doesn't have environment '%s'",
+                                                 workspace.getName(),
+                                                 envName));
+        }
+        if (environment.getRecipe() != null && !"docker".equals(environment.getRecipe().getType())) {
```

Comment:
Move "docker" hardcoded string to some shared constant, I'm sure that other components may have checking like this. It will help us with refactoring in future.